pkg/providers/vsphere/vsphere.go

```diff
@@ -1148,15 +1156,68 @@ func (p *vsphereProvider) InstallCustomProviderComponents(ctx context.Context, k
 	return nil
 }
 
-func (p *vsphereProvider) PostBootstrapDeleteForUpgrade(ctx context.Context) error {
+// PostBootstrapDeleteForUpgrade runs any provider-specific operations after bootstrap cluster has been deleted.
+func (p *vsphereProvider) PostBootstrapDeleteForUpgrade(ctx context.Context, cluster *types.Cluster) error {
+	// Delete the daemonset that was used to disable udp offloading in ubuntu/redhat nodes.
+	logger.V(3).Info("Deleting vsphere-disable-udp-offload daemonset")
+	if err := p.providerKubectlClient.DeleteDaemonSet(ctx, ethtoolDaemonSetName, constants.EksaSystemNamespace, cluster); err != nil {
+		return fmt.Errorf("deleting daemonset: %v", err)
+	}
 	return nil
 }
 
-// PreCoreComponentsUpgrade staisfies the Provider interface.
+// PreCoreComponentsUpgrade satisfies the Provider interface.
 func (p *vsphereProvider) PreCoreComponentsUpgrade(
 	ctx context.Context,
 	cluster *types.Cluster,
 	clusterSpec *cluster.Spec,
 ) error {
+	// This is only needed for EKS-A v0.17 because UDP offloading needs to be disabled for
+	// the new Cilium version to work with VSphere clusters. Since our templates that were shipped
+	// in v0.16 releases still have UDP offloading enabled in the nodes, we must apply the daemon set
+	// to disable UDP offloading in all the nodes before upgrading Cilium.
+	// We will remove this in EKS-A release v0.18.0.
+	return p.applyEthtoolDaemonSet(ctx, cluster, clusterSpec)
+}
+
+func (p *vsphereProvider) applyEthtoolDaemonSet(ctx context.Context, cluster *types.Cluster, clusterSpec *cluster.Spec) error {
+	for _, mc := range clusterSpec.Config.VSphereMachineConfigs {
+		// This is only needed for Ubuntu and Redhat OS.
+		// We only need to check one since the OS family has to be the same for all machine configs
+		if mc.Spec.OSFamily == v1alpha1.Bottlerocket {
+			return nil
+		}
+	}
+	logger.V(4).Info("Applying vsphere-disable-udp-offload daemonset")
+	bundle := clusterSpec.VersionsBundle
+	values := map[string]interface{}{
+		"eksaSystemNamespace": constants.EksaSystemNamespace,
+		"kindNodeImage":       bundle.EksD.KindNode.VersionedImage(),
+	}
+	b, err := templater.Execute(ethtoolDaemonSetObject, values)
+	if err != nil {
+		return err
+	}
+
+	if err = p.providerKubectlClient.ApplyKubeSpecFromBytes(ctx, cluster, b); err != nil {
+		return err
+	}
+
+	return p.Retrier.Retry(
+		func() error {
+			return p.checkEthtoolDaemonSetReady(ctx, cluster)
+		},
+	)
+}
+
+func (p *vsphereProvider) checkEthtoolDaemonSetReady(ctx context.Context, cluster *types.Cluster) error {
+	ethtoolDaemonSet, err := p.providerKubectlClient.GetDaemonSet(ctx, ethtoolDaemonSetName, constants.EksaSystemNamespace, cluster.KubeconfigFile)
+	if err != nil {
+		return err
+	}
+
+	if ethtoolDaemonSet.Status.DesiredNumberScheduled != ethtoolDaemonSet.Status.NumberReady {
```

Comment:
won't the pods be ready before the command finishes?

Reply:
Talked offline. Added a readiness probe to the container